File: learnLeetCode_remote/Day15/Part3/min_refuel_stops.cpp
#include <bits/stdc++.h>
using namespace std;

// LeetCode 871 — Minimum Number of Refueling Stops
// Greedy with max-heap of reachable stations' fuel.

int minRefuelStops(int target, int startFuel, vector<vector<int>> stations){
    long long fuel = startFuel;
    int ans = 0, i = 0, n = stations.size();
    priority_queue<int> pq;
    while (fuel < target){
        while (i < n && stations[i][0] <= fuel){
            pq.push(stations[i][1]);
            ++i;
        }
        if (pq.empty()) return -1;
        fuel += pq.top(); pq.pop();
        ++ans;
    }
    return ans;
}

int main(){
    cout << minRefuelStops(100, 10, {{10,60},{20,30},{30,30},{60,40}}) << "\n"; // 2
    return 0;
}

File: learnLeetCode/Day15/Part3/min_refuel_stops.py
import heapq
from typing import List

# LeetCode 871 — Minimum Number of Refueling Stops
# Greedy with max-heap of fuel from stations we've passed.

def minRefuelStops(target: int, startFuel: int, stations: List[List[int]]) -> int:
    fuel = startFuel
    ans = 0
    i = 0
    n = len(stations)
    maxheap = []  # store negative for max-heap behavior via heapq
    while fuel < target:
        # push all stations within current reach
        while i < n and stations[i][0] <= fuel:
            heapq.heappush(maxheap, -stations[i][1])
            i += 1
        if not maxheap:
            return -1
        fuel += -heapq.heappop(maxheap)
        ans += 1
    return ans

if __name__ == "__main__":
    print(minRefuelStops(100, 10, [[10,60],[20,30],[30,30],[60,40]]))  # 2
